### PR TITLE
Estimate swap

### PR DESCRIPTION
**Pull request by vuong177**
Closes: #2337
## What is the purpose of the change

Change `EstimateSwap` to not require assets.

## Brief Changelog
- Remove field `Sender` in `EstimateSwap` queries.

## Testing and Verifying

This change is a trivial rework / code cleanup without any test coverage.
## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no) no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no) no 
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented) not applicable

### CODE REVIEW

File: proto/osmosis/gamm/v1beta1/query.proto
Comment (ValarDragon) [inline proto/osmosis/gamm/v1beta1/query.proto:150]: Thanks for catching this, pretty funny mistake lol!

However, please don't change the field numbers in proto files like this, lets keep them at the old field numbers, this makes it so legacy clients don't have to update, to 'just work' with the changed proto.

Comment (vuong177) [reply proto/osmosis/gamm/v1beta1/query.proto:150]: Yep, I'll return old field number.

**Comment by vuong177** [comment]
I've created a PR  https://github.com/osmosis-labs/bindings/pull/37 to remove sender field.

File: wasmbinding/queries.go
Comment (mattverse) [inline wasmbinding/queries.go:103]: Can we change this API as well? We can erase the sender Addr here as well

File: proto/osmosis/gamm/v1beta1/query.proto
Comment (mattverse) [inline proto/osmosis/gamm/v1beta1/query.proto:167]: This change is going to be a breaking change anyhow, Can we update the field numbers? (e.g pool_id -> 1)

File: x/gamm/keeper/swap.go
Comment (mattverse) [inline x/gamm/keeper/swap.go:154]: I don't think calling a setter method(`setPool`) inside a query method is a safe way of doing this IMO.
We should either completely separate out the querying methods from the existing method or create a new separate method for this estimate call

**Review by mattverse — CHANGES_REQUESTED** [review]
Thanks for tackling this! I personally don't think it's a safe way to have a setter inside a query method nor do I think using the same route for swaps and queries is a safe way of doing this. 

I would like to suggest further separation of the setters inside the existing method of swapping, or create a new method for this query

File: proto/osmosis/gamm/v1beta1/query.proto
Comment (mattverse) [reply proto/osmosis/gamm/v1beta1/query.proto:167]: Ah just checked the convo above resolving this

File: x/gamm/keeper/swap.go
Comment (ValarDragon) [reply x/gamm/keeper/swap.go:154]: Yeah, this is concerning, great catch.

So turns out some queries in the SDK already do this, so for now this is grudgingly ok, we should cache the context in our query layer to prevent this for now.

File: x/gamm/keeper/swap.go
Comment (ValarDragon) [inline x/gamm/keeper/swap.go:157]: I don't think we should do this

**Comment by vuong177** [comment]
Thanks @ValarDragon @mattverse. I'll make your suggestion.

File: x/gamm/keeper/swap.go
Comment (vuong177) [reply x/gamm/keeper/swap.go:157]: Yes, I'll revert old logic.

File: x/gamm/keeper/swap.go
Comment (ValarDragon) [reply x/gamm/keeper/swap.go:154]: I suggest we don't block this pr on this

**Review by mattverse — APPROVED** [review]

**Comment by ValarDragon** [comment]
This absolutely needs a changelog update. @mattverse can you PR one?

**Comment by ValarDragon** [comment]
Wait this code completely changed since last review

**Comment by vuong177** [comment]
hey duplicate code beacuse esimate swap have similar logic as real swap. Should we add a boolean variable `estimate` in `UpdatePoolForSwap`

**Comment by vuong177** [comment]
Also add `estimate` in `SwapExactAmountIn` and `MultihopSwapExactAmountIn`

**Comment by mattverse** [comment]
Yeah, with the scope of this PR getting big, the approach I would want to take on re-merging this PR is me shooting a secondary PR on top of this branch for further refactoring and reducing code duplication, after we merge that second PR into this one, we can give it a last review and then re merge this PR.

**Comment by mattverse** [comment]
@vuong177 https://github.com/osmosis-labs/osmosis/pull/2501 would appreciate it if you could also review this :)

**Pull request merged**
